src/metatensor/models/utils/data/writers/xyz.py

```diff
@@ -1,33 +1,114 @@
+from metatensor.torch.atomistic import ModelCapabilities, System
 
 
-def write_xyz(filename: str, predictions: TensorMap, systems: List[System]) -> None:
-    """An ase based xyz file writer
+def write_xyz(
+    filename: str,
+    systems: List[System],
+    capabilities: ModelCapabilities,
+    predictions: Dict[str, TensorMap],
+) -> None:
+    """An ase-based xyz file writer. Writes the systems and predictions to an xyz file.
 
-    :param filename: name of the file to read
-    :param predictions: prediction values written to the file.
-    :param systems: strcutures additional written to the file.
+    :param filename: name of the file to read.
+    :param systems: structures to be written to the file.
+    :param: capabilities: capabilities of the model.
+    :param predictions: prediction values to be written to the file.
     """
-    # Get the target property name:
-    target_name = next(iter(predictions.keys()))
+
+    # we first split the predictions by structure
+    predictions_by_structure: List[Dict[str, TensorMap]] = [{} for _ in systems]
+    split_labels = [
+        Labels(names=["system"], values=torch.tensor([[i_system]]))
+        for i_system in range(len(systems))
+    ]
+    for target_name, target_tensor_map in predictions.items():
+        # split this target by structure
+        split_target = metatensor.torch.split(
+            target_tensor_map, "samples", split_labels
+        )
+        for i_system, system_target in enumerate(split_target):
+            # add the split target to the dict corresponding to the structure
+            predictions_by_structure[i_system][target_name] = system_target
+
+    only_one_energy = (
+        np.sum(
+            [
+                capabilities.outputs[key].quantity == "energy"
+                for key in predictions.keys()
+            ]
+        )
+        == 1
+    )
 
     frames = []
-    for i_system, system in enumerate(systems):
-        info = {
-            target_name: float(predictions[target_name].block().values[i_system, 0])
-        }
+    for system, system_predictions in zip(systems, predictions_by_structure):
+        info = {}
+        arrays = {}
+        for target_name, target_map in system_predictions.items():
+            if len(target_map.keys.names) != 1:
+                raise ValueError(
+                    "Only single-block `TensorMap`s can be "
+                    "written to xyz files for the moment."
+                )
+            block = target_map.block()
+            if block.values.numel() == 1:  # this is a scalar
+                info[target_name] = block.values.item()
+            else:  # this is an array
+                arrays[target_name] = block.values.detach().cpu().numpy()
+            for gradient_name, gradient_block in block.gradients():
+                # here, we assume that gradients are always an array, and never a scalar
+                if capabilities[gradient_name].quantity == "energy":
```

Comment:
These blocks of code are not covered. You can check it from codecov.

https://app.codecov.io/gh/lab-cosmo/metatensor-models/pull/150/blob/src/metatensor/models/utils/data/writers/xyz.py

This code seems to be very similar to what we do for the logger. Is there some merging possible?